toolchain/semantics/semantics_handle_index.cpp

```diff
@@ -27,17 +30,40 @@ auto SemanticsHandleIndexExpression(SemanticsContext& context,
       context.semantics_ir().GetTypeAllowBuiltinTypes(name_node.type_id());
   auto name_type_node = context.semantics_ir().GetNode(name_type_id);
 
-  if (name_type_node.kind() == SemanticsNodeKind::TupleType &&
-      index_node.kind() == SemanticsNodeKind::IntegerLiteral) {
+  if (name_type_node.kind() == SemanticsNodeKind::ArrayType) {
+    auto [bound_id, type_id] = name_type_node.GetAsArrayType();
+    if (index_node.kind() == SemanticsNodeKind::IntegerLiteral) {
+      const auto& index_val = context.semantics_ir().GetIntegerLiteral(
+          index_node.GetAsIntegerLiteral());
+      if (index_val.uge(context.semantics_ir().GetArrayBoundValue(bound_id))) {
+        context.emitter().Emit(
+            parse_node, OutOfBoundsAccess,
+            llvm::APSInt(index_val, /*isUnsigned=*/true),
+            context.semantics_ir().StringifyType(name_node.type_id()));
+      } else {
+        context.AddNodeAndPush(
+            parse_node, SemanticsNode::ArrayIndex::Make(
+                            parse_node, type_id, name_node_id, index_node_id));
+        return true;
+      }
+    } else if (context.ImplicitAsRequired(
+                   index_node.parse_node(), index_node_id,
+                   context.CanonicalizeType(
+                       SemanticsNodeId::BuiltinIntegerType)) !=
+               SemanticsNodeId::BuiltinError) {
+      context.AddNodeAndPush(
+          parse_node, SemanticsNode::ArrayIndex::Make(
+                          parse_node, type_id, name_node_id, index_node_id));
+      return true;
+    }
+  } else if (name_type_node.kind() == SemanticsNodeKind::TupleType &&
+             index_node.kind() == SemanticsNodeKind::IntegerLiteral) {
```

Comment:
As you add ArrayType support, it looks like line 78 (checking IntegerLiteral) is actually specific to TupleType. Can you restructure this `if` statement to group TupleType-specific logic similarly to how you do ArrayType-specific logic?